front/lib/resources/storage/models/runs.ts

```diff
@@ -0,0 +1,129 @@
+import type {
+  CreationOptional,
+  ForeignKey,
+  InferAttributes,
+  InferCreationAttributes,
+  NonAttribute,
+} from "sequelize";
+import { DataTypes, Model } from "sequelize";
+
+import { App } from "@app/lib/models/apps";
+import { Workspace } from "@app/lib/models/workspace";
+import { frontSequelize } from "@app/lib/resources/storage";
+
+export class RunModel extends Model<
+  InferAttributes<RunModel>,
+  InferCreationAttributes<RunModel>
+> {
+  declare id: CreationOptional<number>;
+  declare createdAt: CreationOptional<Date>;
+  declare updatedAt: CreationOptional<Date>;
+
+  declare dustRunId: string;
+  declare runType: string;
+
+  declare appId: ForeignKey<App["id"]>;
+  declare workspaceId: ForeignKey<Workspace["id"]>;
+
+  declare app: NonAttribute<App>;
+}
+
+RunModel.init(
+  {
+    id: {
+      type: DataTypes.INTEGER,
+      autoIncrement: true,
+      primaryKey: true,
+    },
+    createdAt: {
+      type: DataTypes.DATE,
+      allowNull: false,
+      defaultValue: DataTypes.NOW,
+    },
+    updatedAt: {
+      type: DataTypes.DATE,
+      allowNull: false,
+      defaultValue: DataTypes.NOW,
+    },
+    dustRunId: {
+      type: DataTypes.STRING,
+      allowNull: false,
+    },
+    runType: {
+      type: DataTypes.STRING,
+      allowNull: false,
+    },
+  },
+  {
+    modelName: "run",
+    sequelize: frontSequelize,
+    indexes: [
+      { fields: ["workspaceId", "appId", "runType", "createdAt"] },
+      { fields: ["workspaceId", "createdAt"] },
+      { unique: true, fields: ["dustRunId"] },
+    ],
+  }
+);
+App.hasMany(RunModel, {
+  foreignKey: { allowNull: false },
+  onDelete: "CASCADE",
+});
+RunModel.belongsTo(App, {
+  as: "app",
+  foreignKey: { name: "appId", allowNull: false },
+});
+Workspace.hasMany(RunModel, {
+  foreignKey: { allowNull: false },
+  onDelete: "CASCADE",
+});
+
+export class RunUsageModel extends Model<
+  InferAttributes<RunUsageModel>,
+  InferCreationAttributes<RunUsageModel>
+> {
+  declare id: CreationOptional<number>;
+
+  declare runId: ForeignKey<RunModel["id"]>;
+
+  declare providerId: string; //ModelProviderIdType;
+  declare modelId: string; //ModelIdType;
+
+  declare promptTokens: number;
+  declare completionTokens: number;
+}
+
+RunUsageModel.init(
+  {
+    id: {
+      type: DataTypes.INTEGER,
+      autoIncrement: true,
+      primaryKey: true,
+    },
+    providerId: {
+      type: DataTypes.STRING,
+      allowNull: false,
+    },
+    modelId: {
+      type: DataTypes.STRING,
+      allowNull: false,
+    },
+    promptTokens: {
+      type: DataTypes.INTEGER,
+      allowNull: false,
+    },
+    completionTokens: {
+      type: DataTypes.INTEGER,
+      allowNull: false,
+    },
+  },
+  {
+    modelName: "run_usages",
+    sequelize: frontSequelize,
+    indexes: [{ fields: ["runId"] }, { fields: ["providerId", "modelId"] }],
+  }
+);
+
+RunModel.hasMany(RunUsageModel, {
+  foreignKey: { allowNull: false },
+  onDelete: "CASCADE",
```

Comment:
@tdraier we also avoid the `onDelete: "CASCADE"` at it can trigger a lot of unbounded number of sql deletion. Instead, we will implement a proper deletion path in the app to first delete `runs_usage` and then delete `runs`.

Reply:
I added a TODO in the code to implement this behavior 😉.

Reply:
It makes sense on some relations, but in that specific case, a single Run has just a few RunUsages linked - so I don't know if it's worth implement a specifc deletion code .. ?

Reply:
If you only consider a single run scope indeed, but there is an existing code path today that deletes all runs associated to an app. This relation basically doubles the number of deletions.